tfhe/src/core_crypto/entities/modulus_switched_lwe_ciphertext.rs

```diff
@@ -0,0 +1,143 @@
+
+        let ms = |x| decomposer.closest_representable(x) >> (64 - log_modulus);
+
+        for i in 0..len as u64 {
+            let mut j = 64 * i / log_modulus;
+
+            let mut value = ms(slice[j as usize]) >> (i * 64 - j * log_modulus);
+            j += 1;
+
+            while j * log_modulus < ((i + 1) * 64) && j < lwe_size as u64 {
+                value |= ms(slice[j as usize]) << (j * log_modulus - i * 64);
+                j += 1;
+            }
+
+            packed_coeffs[i as usize] = value;
+        }
+
+        Self {
+            packed_coeffs,
+            lwe_dimension: ct.lwe_size().to_lwe_dimension(),
+            log_modulus,
+            ciphertext_modulus: ct.ciphertext_modulus(),
+        }
+    }
+
+    pub fn extract(&self) -> LweCiphertextOwned<u64> {
+        assert!(self.log_modulus <= 64);
+
+        // log_modulus lowest bits et to 1
+        let mask = u64::MAX >> (64 - self.log_modulus);
+
+        let container = (0..(self.lwe_dimension.0 as u64 + 1))
+            .map(|i| {
+                let start = i * self.log_modulus;
+                let end = (i + 1) * self.log_modulus;
+
+                let start_block = start / 64;
+                let start_remainder = start % 64;
+
+                let end_block_inclusive = (end - 1) / 64;
+
+                if start_block == end_block_inclusive {
+                    self.packed_coeffs[start_block as usize] >> start_remainder
+                } else {
+                    assert_eq!(end_block_inclusive, start_block + 1);
+
+                    let first_part = self.packed_coeffs[start_block as usize] >> start_remainder;
+
+                    let second_part = self.packed_coeffs[end_block_inclusive as usize]
+                        << ((64 - start_remainder) % 64);
+
+                    first_part | second_part
+                }
+            })
+            .map(|a| (a & mask) << (64 - self.log_modulus))
+            .collect();
+
+        LweCiphertextOwned::from_container(container, self.ciphertext_modulus)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::core_crypto::prelude::*;
+    use concrete_csprng::seeders::Seed;
+
+    #[test]
+    fn ms_compression_() {
+        ms_compression(33, 10);
+
+        ms_compression(1, 100);
+
+        ms_compression(53, 37);
+
+        ms_compression(63, 63);
+
+        ms_compression(10, 64);
+
+        ms_compression(11, 700);
+
+        ms_compression(12, 751);
+    }
+
+    fn ms_compression(log_modulus: u64, len: usize) {
+        let ciphertext_modulus = CiphertextModulus::new_native();
+
+        let mut seeder = new_seeder();
+
+        let mut encryption_generator =
+            EncryptionRandomGenerator::<ActivatedRandomGenerator>::new(Seed(0), seeder.as_mut());
+
+        let mut lwe = vec![0; len];
+
+        encryption_generator.fill_slice_with_random_uniform_mask(&mut lwe);
+
+        let lwe = LweCiphertextOwned::from_container(lwe, ciphertext_modulus);
```

Comment:
I think I would try to do an actual encryption and check we can encrypt, pack, decrypt and recover the input ? wdyt ?

Could be done in the test folder to also cover more parameter sets with the create parameterized macro

Reply:
I looks to me that the test I wrote is more fine grained, so better
Why do you think a test with encryption and decryption would be a good test?

Reply:
because this way we can make sure we are properly preserving the data

as for the other parameter sets this way we make sure we have support of power of 2 moduli which are essentially the same code

Reply:
Sure, we can do that
Do you mean for non powers of 2 ct_modulus_log?

Reply:
No, non power of 2 moduli are not currently supported in a PBS so we would not be able to refresh the ciphertexts

Reply:
Added in doctest